shared/lib/delegation/utils.ts

```diff
@@ -0,0 +1,244 @@
+ * @returns The concatenated hex string
+ */
+export function concat(values: readonly Hex[]): Hex {
+  return `0x${(values as Hex[]).reduce((acc, x) => acc + remove0x(x), '')}`;
+}
+
+type IsAddressOptions = {
+  /**
+   * Enables strict mode. Whether or not to compare the address against its checksum.
+   *
+   * @default true
+   */
+  strict?: boolean | undefined;
+};
+
+/**
+ * Checks if the given string is a valid address
+ *
+ * @param address - The address to check
+ * @param options - Options for the function
+ * @param options.strict - Enables strict mode. Whether or not to compare the address against its checksum. Defaults to `true`
+ * @returns `true` if the address is valid, `false` otherwise
+ */
+export function isAddress(
+  address: string,
+  options?: IsAddressOptions | undefined,
+): address is Address {
+  const { strict = true } = options || {};
+  const addr = address as Hex;
+  return strict ? isValidChecksumAddress(addr) : isValidHexAddress(addr);
+}
+
+function normalizeSignature(signature: string): string {
+  let active = true;
+  let current = '';
+  let level = 0;
+  let result = '';
+  let valid = false;
+
+  for (let i = 0; i < signature.length; i++) {
+    const char = signature[i];
+
+    // If the character is a separator, we want to reactivate.
+    if (['(', ')', ','].includes(char)) {
+      active = true;
+    }
+
+    // If the character is a "level" token, we want to increment/decrement.
+    if (char === '(') {
+      level++;
+    }
+    if (char === ')') {
+      level--;
+    }
+
+    // If we aren't active, we don't want to mutate the result.
+    if (!active) {
+      continue;
+    }
+
+    // If level === 0, we are at the definition level.
+    if (level === 0) {
+      if (char === ' ' && ['event', 'function', ''].includes(result)) {
+        result = '';
+      } else {
+        result += char;
+
+        // If we are at the end of the definition, we must be finished.
+        if (char === ')') {
+          valid = true;
+          break;
+        }
+      }
+
+      continue;
+    }
+
+    // Ignore spaces
+    if (char === ' ') {
+      // If the previous character is a separator, and the current section isn't empty, we want to deactivate.
+      if (signature[i - 1] !== ',' && current !== ',' && current !== ',(') {
+        current = '';
+        active = false;
+      }
+      continue;
+    }
+
+    result += char;
+    current += char;
+  }
+
+  if (!valid) {
+    throw new Error('Unable to normalize signature.');
+  }
+
+  return result;
+}
+
+function hashSignature(fn: string) {
+  return keccak(Buffer.from(stringToBytes(normalizeSignature(fn))));
+}
+/**
+ * Returns the function selector for a given function definition.
+ *
+ * @param fn
+ * @example
+ * const selector = toFunctionSelector('function ownerOf(uint256 tokenId)')
+ * // 0x6352211e
+ */
+export const toFunctionSelector = (fn: string): Hex => {
+  const hash = toHex(hashSignature(fn).subarray(0, 10));
```

Comment:
So, uh, I may have gotten this wrong on my first suggestion(sorry).


Should it be:

```suggestion
  const hash = toHex(hashSignature(fn).subarray(0, 5));
```

?

Reply:
Yup! It's actually `4` bytes. Just pushed a change.

Reply:
ah, right! the `toHex` adds in two additional chracters, hence the original `10`. thanks!